File: engines/bmx_plus.py
"""
BMX+ — Entropy-Weighted Lexical Search via Term-At-A-Time Evaluation.

Successor to BMX (arXiv:2408.06643). Builds on BM25's proven TF saturation
curve with three innovations:
  1. Term-adaptive entropy-aware IDF (γₜ = IDFₜ / IDF_max)
  2. Variance-blended informativeness (Shannon ↔ IDF, smooth transition)
  3. tanh Soft-AND coverage bonus (RankEvolve-inspired, anti-dominance)

All executed within a TAAT posting-list architecture for 3.4–30× speedup.

Score per document:

  score(D, Q) = Σᵢ [eIDF(qᵢ) · TF_BM25(qᵢ, D) · qtf(qᵢ)]
              + [Σᵢ tanh(scoreᵢ) / |Q|] · Σᵢ [γₜᵢ · info(qᵢ) · qtf(qᵢ)]

  eIDF(q)     = IDF(q) · (1 + γₜ · info(q))        — entropy-aware IDF
  γₜ          = IDF(t) / IDF_max                     — term-adaptive weight
  TF_BM25     = tf·(k₁+1) / (tf + k₁·(1−b+b·dl/avgdl))  — standard BM25 TF
  info(q)     = blend(shannon_info, idf_info, variance)    — variance-blended

  k₁ = 1.5, b = 0.75                                — proven BM25 defaults
  IDF_max = max(IDF(t) for all t in corpus)          — normalisation anchor

All parameters derived from corpus statistics — no manual tuning required.
Drop-in replacement for BM25Index — identical public API.
"""

import logging
import math
import re
from collections import Counter, defaultdict
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)

_WORD_RE = re.compile(r"\b\w+\b")


def _fast_sigmoid(x: float) -> float:
    """Padé rational approximation to σ(x) = 1/(1+e⁻ˣ). |error| < 0.01."""
    if x >= 8.0:
        return 1.0
    if x <= -8.0:
        return 0.0
    x2 = x * x
    x3 = x2 * x
    return (x3 + 6.0 * x + 12.0) / (x3 + 12.0 * x + 48.0)


@dataclass
class BMXPlusIndex:
    """
    In-memory BMX+ search index with TAAT (Term-At-A-Time) scoring.

    Matches BM25 ranking quality (−0.15% avg NDCG@10) with superior recall
    (+0.10% avg Recall@100) at 3.4–30× throughput on BEIR benchmarks.

    Self-tuning — all parameters derived from corpus statistics:
      γₜ = IDF(t) / IDF_max  — term-adaptive entropy weight
      k₁ = 1.5, b = 0.75    — standard BM25 TF parameters

    Optional overrides:
      alpha_override: Force α value (None = auto)
      beta_override:  Force β value (None = auto)
    """

    alpha_override: float | None = None
    beta_override: float | None = None
    normalize_scores: bool = False

    # ── Document storage ──
    _documents: dict[str, list[str]] = field(default_factory=dict)
    _doc_lengths: dict[str, int] = field(default_factory=dict)
    _avg_doc_length: float = 0.0
    _total_docs: int = 0
    _is_built: bool = False

    # ── Posting lists: term → {chunk_id: tf} ──
    _posting_lists: dict[str, dict[str, int]] = field(default_factory=dict)

    # ── Frequencies and IDF ──
    _doc_freqs: dict[str, int] = field(default_factory=dict)
    _idf_cache: dict[str, float] = field(default_factory=dict)

    # ── Entropy state ──
    _term_total_freqs: dict[str, int] = field(default_factory=dict)
    _term_entropy: dict[str, float] = field(default_factory=dict)
    _term_info: dict[str, float] = field(default_factory=dict)
    _dirty_terms: set[str] = field(default_factory=set)

    # ── Self-tuning parameters ──
    _alpha: float = 1.0
    _beta: float = 0.01
    _idf_max: float = 1.0

    def __post_init__(self):
        self._documents = {}
        self._doc_lengths = {}
        self._posting_lists = {}
        self._doc_freqs = {}
        self._idf_cache = {}
        self._term_total_freqs = {}
        self._term_entropy = {}
        self._term_info = {}
        self._dirty_terms = set()

    # ════════════════════════════════════════════════════════════════════
    #  Tokenisation
    # ════════════════════════════════════════════════════════════════════

    @staticmethod
    def _tokenize(text: str) -> list[str]:
        return _WORD_RE.findall(text.lower())

    # ════════════════════════════════════════════════════════════════════
    #  Self-Tuning Parameters
    # ════════════════════════════════════════════════════════════════════

    def _compute_parameters(self):
        N = self._total_docs
        avgdl = self._avg_doc_length

        self._alpha = (
            self.alpha_override
            if self.alpha_override is not None
            else max(0.5, min(1.5, avgdl / 100.0))
        )
        self._beta = (
            self.beta_override
            if self.beta_override is not None
            else (1.0 / math.log(1.0 + N) if N > 0 else 0.01)
        )
        # Compute IDF_max for term-adaptive scaling: γₜ = IDFₜ / IDF_max
        # Rare terms get full entropy weight, common terms get none,
        # independent of corpus size.
        self._idf_max = max(self._idf_cache.values()) if self._idf_cache else 1.0

    # ════════════════════════════════════════════════════════════════════
    #  Entropy Computation
    # ════════════════════════════════════════════════════════════════════

    def _compute_idf(self, term: str) -> float:
        """Lucene-variant BM25 IDF (always non-negative)."""
        df = self._doc_freqs.get(term, 0)
        N = self._total_docs
        if df > 0 and N > 0:
            return math.log((N - df + 0.5) / (df + 0.5) + 1.0)
        return 0.0

    def _compute_term_entropies(self, terms: set[str] | None = None):
        """Compute term informativeness via smoothly blended entropy.

        Blends Shannon entropy (when TFs vary across documents) with
        IDF-derived informativeness (1 - df/N) using TF variance as
        the interpolation weight. This avoids a hard discontinuity
        when a single document with tf=2 flips the computation regime.
        """
        target = terms if terms is not None else set(self._doc_freqs.keys())
        N = self._total_docs

        for term in target:
            df = self._doc_freqs.get(term, 0)
            self._idf_cache[term] = self._compute_idf(term)

            if df < 2:
                self._term_entropy[term] = 0.0
                self._term_info[term] = 1.0
                continue

            posting = self._posting_lists.get(term, {})
            if not posting:
                self._term_entropy[term] = 1.0
                self._term_info[term] = 0.0
                continue

            # IDF-derived informativeness (always available)
            idf_info = 1.0 - df / N if N > 0 else 0.0

            # Compute TF variance to determine blend weight
            tf_vals = list(posting.values())
            n_post = len(tf_vals)
            mean_tf = sum(tf_vals) / n_post
            variance = sum((v - mean_tf) ** 2 for v in tf_vals) / n_post

            # Smooth blend: α→0 when uniform, α→1 when varied
            # epsilon=1.0 gives a gentle sigmoid-like transition
            blend_alpha = variance / (variance + 1.0)

            if blend_alpha < 0.001:
                # Essentially uniform — pure IDF-derived info
                self._term_info[term] = idf_info
                self._term_entropy[term] = 1.0 - idf_info
            else:
                # Compute Shannon entropy of sigmoid-mapped distribution
                mapped = [_fast_sigmoid(float(tf)) for tf in tf_vals]
                total_mapped = sum(mapped)
                if total_mapped == 0.0:
                    self._term_info[term] = idf_info
                    self._term_entropy[term] = 1.0 - idf_info
                    continue

                entropy = 0.0
                inv_total = 1.0 / total_mapped
                for m_val in mapped:
                    p = m_val * inv_total
                    if p > 0.0:
                        entropy -= p * math.log(p)

                max_ent = math.log(df)
                norm_ent = min(entropy / max_ent, 1.0) if max_ent > 0.0 else 0.0
                shannon_info = max(1.0 - norm_ent, 0.0)

                # Blend: Shannon when TFs vary, IDF-derived when uniform
                blended = blend_alpha * shannon_info + (1.0 - blend_alpha) * idf_info
                self._term_info[term] = blended
                self._term_entropy[term] = 1.0 - blended

    def _flush_dirty_entropies(self, query_terms: set[str]):
        """Lazily recompute entropies only for dirty terms in the query."""
        if not self._dirty_terms:
            return
        to_flush = self._dirty_terms & query_terms
        if to_flush:
            self._compute_term_entropies(to_flush)
            self._dirty_terms -= to_flush

    # ════════════════════════════════════════════════════════════════════
    #  Build Index
    # ════════════════════════════════════════════════════════════════════

    def _reset(self):
        """Clear all index state."""
        self._documents.clear()
        self._doc_lengths.clear()
        self._posting_lists.clear()
        self._doc_freqs.clear()
        self._idf_cache.clear()
        self._term_total_freqs.clear()
        self._term_entropy.clear()
        self._term_info.clear()
        self._dirty_terms.clear()
        self._avg_doc_length = 0.0
        self._total_docs = 0
        self._is_built = False

    def build_index(self, chunks: list[dict] | None = None):
        """Build from [{"chunk_id": str, "text": str}, ...]."""
        if chunks:
            self._reset()
            for chunk in chunks:
                tokens = self._tokenize(chunk["text"])
                cid = chunk["chunk_id"]
                self._documents[cid] = tokens
                self._doc_lengths[cid] = len(tokens)

        N = len(self._documents)
        if N == 0:
            self._is_built = True
            return

        self._total_docs = N
        self._avg_doc_length = sum(self._doc_lengths.values()) / N

        # Build posting lists, document frequencies, total term frequencies
        for cid, tokens in self._documents.items():
            term_counts = Counter(tokens)
            for term, count in term_counts.items():
                if term not in self._posting_lists:
                    self._posting_lists[term] = {}
                self._posting_lists[term][cid] = count

                self._doc_freqs[term] = self._doc_freqs.get(term, 0) + 1
                self._term_total_freqs[term] = (
                    self._term_total_freqs.get(term, 0) + count
                )

        self._compute_term_entropies()  # also populates _idf_cache
        self._compute_parameters()       # uses _idf_cache for _idf_max
        self._is_built = True

        logger.debug(
            "BMX+ index built: %d docs, %d terms, a=%.3f b=%.4f idf_max=%.4f",
            N, len(self._posting_lists), self._alpha, self._beta, self._idf_max,
        )

    # ════════════════════════════════════════════════════════════════════
    #  TAAT Search
    # ════════════════════════════════════════════════════════════════════

    def search(self, query: str, top_k: int = 10) -> list[tuple[str, float]]:
        """Search via Term-At-A-Time posting list traversal.

        Uses BM25's proven TF saturation curve inside the TAAT architecture,
        with term-adaptive entropy-aware IDF and tanh Soft-AND coverage bonus.
        """
        if not self._is_built or not self._documents:
            return []

        query_tokens = self._tokenize(query)
        if not query_tokens:
            return []

        query_tf = Counter(query_tokens)
        unique_query = set(query_tokens)
        m = len(unique_query)

        self._flush_dirty_entropies(unique_query)

        # ── Cache locals for the hot loop ──
        k1 = 1.5
        b = 0.75
        idf_max = self._idf_max
        avgdl = self._avg_doc_length
        inv_idf_max = 1.0 / idf_max if idf_max > 0.0 else 1.0
        doc_lengths = self._doc_lengths
        posting_lists = self._posting_lists
        idf_cache = self._idf_cache
        term_info = self._term_info
        inv_m = 1.0 / m
        _tanh = math.tanh

        # Per-query informativeness weights
        info_weights: dict[str, float] = {}
        info_total = 0.0
        for t in unique_query:
            v = term_info.get(t, 0.0)
            info_weights[t] = v
            info_total += v

        # ── TAAT accumulation ──
        scores: dict[str, float] = defaultdict(float)
        info_accum: dict[str, float] = defaultdict(float)
        tanh_coverage: dict[str, float] = defaultdict(float)

        for term, q_tf in query_tf.items():
            posting = posting_lists.get(term)
            if not posting:
                continue

            idf = idf_cache.get(term, 0.0)
            info_qi = info_weights.get(term, 0.0)

            # Term-adaptive γₜ: rare terms get full entropy weight,
            # common terms get none — corpus-size independent
            gamma_t = idf * inv_idf_max
            info_x_q = gamma_t * info_qi * q_tf

            if idf <= 0.0:
                for cid in posting:
                    tanh_coverage[cid] += 1.0
                    info_accum[cid] += info_x_q
                continue

            # Entropy-aware IDF with term-adaptive scaling
            eidf = idf * (1.0 + gamma_t * info_qi)

            for cid, tf in posting.items():
                # BM25 TF saturation — proven, well-calibrated
                dl = doc_lengths[cid]
                tf_sat = (tf * (k1 + 1.0)) / (tf + k1 * (1.0 - b + b * dl / avgdl))

                term_score = eidf * tf_sat * q_tf
                scores[cid] += term_score
                info_accum[cid] += info_x_q
                tanh_coverage[cid] += _tanh(term_score)

        if not scores:
            return []

        # ── Final: Soft-AND coverage bonus ──
        final: dict[str, float] = {}
        for cid, base in scores.items():
            soft_and = tanh_coverage[cid] * inv_m
            final[cid] = base + soft_and * info_accum[cid]

        sorted_results = sorted(final.items(), key=lambda x: x[1], reverse=True)

        if self.normalize_scores and sorted_results:
            max_score = sorted_results[0][1]
            if max_score > 0.0:
                sorted_results = [(c, s / max_score) for c, s in sorted_results]

        return sorted_results[:top_k]

    # ════════════════════════════════════════════════════════════════════
    #  Incremental Updates
    # ════════════════════════════════════════════════════════════════════

    def update_index(self, chunk_id: str, text: str):
        """Add or replace a document (lazy entropy recomputation)."""
        if chunk_id in self._documents:
            self.remove_from_index(chunk_id)

        tokens = self._tokenize(text)
        self._documents[chunk_id] = tokens
        self._doc_lengths[chunk_id] = len(tokens)

        term_counts = Counter(tokens)
        affected: set[str] = set()

        for term, count in term_counts.items():
            if term not in self._posting_lists:
                self._posting_lists[term] = {}
            self._posting_lists[term][chunk_id] = count

            self._doc_freqs[term] = self._doc_freqs.get(term, 0) + 1
            self._term_total_freqs[term] = (
                self._term_total_freqs.get(term, 0) + count
            )
            affected.add(term)

        self._total_docs = len(self._documents)
        self._avg_doc_length = sum(self._doc_lengths.values()) / self._total_docs
        self._compute_parameters()
        self._dirty_terms.update(affected)

    def remove_from_index(self, chunk_id: str):
        """Remove a document (lazy entropy recomputation)."""
        if chunk_id not in self._documents:
            return

        tokens = self._documents[chunk_id]
        term_counts = Counter(tokens)
        affected: set[str] = set()

        for term, count in term_counts.items():
            posting = self._posting_lists.get(term)
            if posting and chunk_id in posting:
                del posting[chunk_id]
                if not posting:
                    del self._posting_lists[term]

            self._doc_freqs[term] = max(self._doc_freqs.get(term, 1) - 1, 0)
            if self._doc_freqs[term] == 0:
                self._doc_freqs.pop(term, None)
                self._idf_cache.pop(term, None)
                self._term_entropy.pop(term, None)
                self._term_info.pop(term, None)

            self._term_total_freqs[term] = max(
                self._term_total_freqs.get(term, count) - count, 0
            )
            if self._term_total_freqs[term] == 0:
                self._term_total_freqs.pop(term, None)

            affected.add(term)

        del self._documents[chunk_id]
        del self._doc_lengths[chunk_id]

        self._total_docs = len(self._documents)
        if self._total_docs > 0:
            self._avg_doc_length = (
                sum(self._doc_lengths.values()) / self._total_docs
            )
        else:
            self._avg_doc_length = 0.0
        self._compute_parameters()
        self._dirty_terms.update(affected)

    # ════════════════════════════════════════════════════════════════════
    #  Properties
    # ════════════════════════════════════════════════════════════════════

    @property
    def document_count(self) -> int:
        return self._total_docs

    @property
    def vocabulary_size(self) -> int:
        return len(self._posting_lists)

    def get_stats(self) -> dict:
        return {
            "total_docs": self._total_docs,
            "vocabulary_size": self.vocabulary_size,
            "avg_doc_length": round(self._avg_doc_length, 2),
            "alpha": round(self._alpha, 4),
            "beta": round(self._beta, 4),
            "idf_max": round(self._idf_max, 4),
        }

File: src/bmx-plus.ts
/**
 * BMX+ — Entropy-Weighted Lexical Search via Term-At-A-Time Evaluation.
 *
 * Successor to BMX (arXiv:2408.06643). Builds on BM25's proven TF saturation
 * curve with three innovations:
 *   1. Term-adaptive entropy-aware IDF (γt = IDFt / IDF_max)
 *   2. Variance-blended informativeness (Shannon ↔ IDF, smooth transition)
 *   3. tanh Soft-AND coverage bonus (RankEvolve-inspired, anti-dominance)
 *
 * All executed within a TAAT posting-list architecture for 3.4–30× speedup.
 *
 * For the source route this is the lexical RELEVANCE engine: build the index
 * once over code blocks (chunk_id = block/line-range id), search once with a
 * scan query, read the ranking. The index is build-once / read-once — the
 * incremental streaming API (updateIndex/removeFromIndex + lazy dirty-term
 * recomputation) and the unused alpha/beta self-tuning knobs were removed;
 * neither ever fed search().
 */

// Module-level word regex
// JS \w is not unicode-aware — use \p{L} and \p{N} for equivalent behavior
// with the 'u' flag to match \w's unicode behavior.
const _WORD_RE = /[\p{L}\p{N}_]+/gu;

/**
 * Padé rational approximation to σ(x) = 1/(1+e^-x). |error| < 0.01.
 */
function _fastSigmoid(x: number): number {
  if (x >= 8.0) return 1.0;
  if (x <= -8.0) return 0.0;
  const x2 = x * x;
  const x3 = x2 * x;
  return (x3 + 6.0 * x + 12.0) / (x3 + 12.0 * x + 48.0);
}

interface Chunk {
  chunk_id: string;
  text: string;
}

export class BMXPlusIndex {
  // ── Public configuration ──
  readonly normalizeScores: boolean;

  // ── Document storage ──
  private _documents: Map<string, string[]>;
  private _docLengths: Map<string, number>;
  private _avgDocLength: number;
  private _totalDocs: number;
  private _isBuilt: boolean;

  // ── Posting lists: term → {chunk_id: tf} ──
  private _postingLists: Map<string, Map<string, number>>;

  // ── Frequencies and IDF ──
  private _docFreqs: Map<string, number>;
  private _idfCache: Map<string, number>;

  // ── Entropy state ──
  private _termTotalFreqs: Map<string, number>;
  private _termEntropy: Map<string, number>;
  private _termInfo: Map<string, number>;

  // ── Term-adaptive scaling ──
  private _idfMax: number;

  constructor(normalizeScores: boolean = false) {
    this.normalizeScores = normalizeScores;

    this._documents = new Map();
    this._docLengths = new Map();
    this._avgDocLength = 0.0;
    this._totalDocs = 0;
    this._isBuilt = false;

    this._postingLists = new Map();
    this._docFreqs = new Map();
    this._idfCache = new Map();

    this._termTotalFreqs = new Map();
    this._termEntropy = new Map();
    this._termInfo = new Map();

    this._idfMax = 1.0;
  }

  // ════════════════════════════════════════════════════════════════════
  //  Tokenisation
  // ════════════════════════════════════════════════════════════════════

  /**
   * JS /g regex is stateful — use String.prototype.match which returns all
   * matches without the statefulness issue of repeated .test()/.exec() calls.
   */
  static _tokenize(text: string): string[] {
    // String.match with /g returns all matches as string[] or null
    // We use the unicode-aware regex defined above
    const lower = text.toLowerCase();
    const matches = lower.match(_WORD_RE);
    return matches !== null ? matches : [];
  }

  // ════════════════════════════════════════════════════════════════════
  //  Term-Adaptive Scaling
  // ════════════════════════════════════════════════════════════════════

  private _computeIdfMax(): void {
    // Compute IDF_max for term-adaptive scaling: γt = IDFt / IDF_max
    // Rare terms get full entropy weight, common terms get none,
    // independent of corpus size.
    if (this._idfCache.size > 0) {
      let max = -Infinity;
      for (const v of this._idfCache.values()) {
        if (v > max) max = v;
      }
      this._idfMax = max;
    } else {
      this._idfMax = 1.0;
    }
  }

  // ════════════════════════════════════════════════════════════════════
  //  Entropy Computation
  // ════════════════════════════════════════════════════════════════════

  /** Lucene-variant BM25 IDF (always non-negative). */
  private _computeIdf(term: string): number {
    const df = this._docFreqs.get(term) ?? 0;
    const N = this._totalDocs;
    if (df > 0 && N > 0) {
      return Math.log((N - df + 0.5) / (df + 0.5) + 1.0);
    }
    return 0.0;
  }

  /**
   * Compute term informativeness via smoothly blended entropy.
   *
   * Blends Shannon entropy (when TFs vary across documents) with
   * IDF-derived informativeness (1 - df/N) using TF variance as
   * the interpolation weight. This avoids a hard discontinuity
   * when a single document with tf=2 flips the computation regime.
   */
  private _computeTermEntropies(terms?: Set<string>): void {
    const target: Set<string> =
      terms !== undefined ? terms : new Set(this._docFreqs.keys());
    const N = this._totalDocs;

    for (const term of target) {
      const df = this._docFreqs.get(term) ?? 0;
      this._idfCache.set(term, this._computeIdf(term));

      if (df < 2) {
        this._termEntropy.set(term, 0.0);
        this._termInfo.set(term, 1.0);
        continue;
      }

      const posting = this._postingLists.get(term);
      if (!posting || posting.size === 0) {
        this._termEntropy.set(term, 1.0);
        this._termInfo.set(term, 0.0);
        continue;
      }

      // IDF-derived informativeness (always available)
      const idfInfo = N > 0 ? 1.0 - df / N : 0.0;

      // Compute TF variance to determine blend weight
      const tfVals: number[] = [...posting.values()];
      const nPost = tfVals.length;
      const meanTf = tfVals.reduce((a, b) => a + b, 0) / nPost;
      const variance =
        tfVals.reduce((acc, v) => acc + (v - meanTf) ** 2, 0) / nPost;

      // Smooth blend: α→0 when uniform, α→1 when varied
      // epsilon=1.0 gives a gentle sigmoid-like transition
      const blendAlpha = variance / (variance + 1.0);

      if (blendAlpha < 0.001) {
        // Essentially uniform — pure IDF-derived info
        this._termInfo.set(term, idfInfo);
        this._termEntropy.set(term, 1.0 - idfInfo);
      } else {
        // Compute Shannon entropy of sigmoid-mapped distribution
        const mapped: number[] = tfVals.map((tf) => _fastSigmoid(tf));
        const totalMapped = mapped.reduce((a, b) => a + b, 0);

        if (totalMapped === 0.0) {
          this._termInfo.set(term, idfInfo);
          this._termEntropy.set(term, 1.0 - idfInfo);
          continue;
        }

        let entropy = 0.0;
        const invTotal = 1.0 / totalMapped;
        for (const mVal of mapped) {
          const p = mVal * invTotal;
          if (p > 0.0) {
            entropy -= p * Math.log(p);
          }
        }

        const maxEnt = Math.log(df);
        const normEnt = maxEnt > 0.0 ? Math.min(entropy / maxEnt, 1.0) : 0.0;
        const shannonInfo = Math.max(1.0 - normEnt, 0.0);

        // Blend: Shannon when TFs vary, IDF-derived when uniform
        const blended = blendAlpha * shannonInfo + (1.0 - blendAlpha) * idfInfo;
        this._termInfo.set(term, blended);
        this._termEntropy.set(term, 1.0 - blended);
      }
    }
  }

  // ════════════════════════════════════════════════════════════════════
  //  Build Index
  // ════════════════════════════════════════════════════════════════════

  /** Clear all index state. */
  _reset(): void {
    this._documents.clear();
    this._docLengths.clear();
    this._postingLists.clear();
    this._docFreqs.clear();
    this._idfCache.clear();
    this._termTotalFreqs.clear();
    this._termEntropy.clear();
    this._termInfo.clear();
    this._avgDocLength = 0.0;
    this._totalDocs = 0;
    this._isBuilt = false;
  }

  /** Build from [{chunk_id: string, text: string}, ...]. */
  buildIndex(chunks?: Chunk[]): void {
    if (chunks && chunks.length > 0) {
      this._reset();
      for (const chunk of chunks) {
        const tokens = BMXPlusIndex._tokenize(chunk.text);
        const cid = chunk.chunk_id;
        this._documents.set(cid, tokens);
        this._docLengths.set(cid, tokens.length);
      }
    }

    const N = this._documents.size;
    if (N === 0) {
      this._isBuilt = true;
      return;
    }

    this._totalDocs = N;
    let totalLen = 0;
    for (const l of this._docLengths.values()) totalLen += l;
    this._avgDocLength = totalLen / N;

    // Build posting lists, document frequencies, total term frequencies
    for (const [cid, tokens] of this._documents.entries()) {
      // Counter equivalent
      const termCounts = new Map<string, number>();
      for (const token of tokens) {
        termCounts.set(token, (termCounts.get(token) ?? 0) + 1);
      }

      for (const [term, count] of termCounts.entries()) {
        if (!this._postingLists.has(term)) {
          this._postingLists.set(term, new Map());
        }
        this._postingLists.get(term)!.set(cid, count);

        this._docFreqs.set(term, (this._docFreqs.get(term) ?? 0) + 1);
        this._termTotalFreqs.set(
          term,
          (this._termTotalFreqs.get(term) ?? 0) + count
        );
      }
    }

    this._computeTermEntropies(); // also populates _idfCache
    this._computeIdfMax();        // uses _idfCache for _idfMax
    this._isBuilt = true;
  }

  // ════════════════════════════════════════════════════════════════════
  //  TAAT Search
  // ════════════════════════════════════════════════════════════════════

  /**
   * Search via Term-At-A-Time posting list traversal.
   *
   * Uses BM25's proven TF saturation curve inside the TAAT architecture,
   * with term-adaptive entropy-aware IDF and tanh Soft-AND coverage bonus.
   */
  search(query: string, topK: number = 10): Array<[string, number]> {
    if (!this._isBuilt || this._documents.size === 0) return [];

    const queryTokens = BMXPlusIndex._tokenize(query);
    if (queryTokens.length === 0) return [];

    // Counter equivalent for query tokens
    const queryTf = new Map<string, number>();
    for (const token of queryTokens) {
      queryTf.set(token, (queryTf.get(token) ?? 0) + 1);
    }
    const uniqueQuery = new Set(queryTokens);
    const m = uniqueQuery.size;

    // Cache locals for the hot loop
    const k1 = 1.5;
    const b = 0.75;
    const idfMax = this._idfMax;
    const avgdl = this._avgDocLength;
    const invIdfMax = idfMax > 0.0 ? 1.0 / idfMax : 1.0;
    const docLengths = this._docLengths;
    const postingLists = this._postingLists;
    const idfCache = this._idfCache;
    const termInfo = this._termInfo;
    const invM = 1.0 / m;

    // Per-query informativeness weights
    const infoWeights = new Map<string, number>();
    let infoTotal = 0.0;
    for (const t of uniqueQuery) {
      const v = termInfo.get(t) ?? 0.0;
      infoWeights.set(t, v);
      infoTotal += v;
    }
    const scores = new Map<string, number>();
    const infoAccum = new Map<string, number>();
    const tanhCoverage = new Map<string, number>();

    for (const [term, qTf] of queryTf.entries()) {
      const posting = postingLists.get(term);
      if (!posting || posting.size === 0) continue;

      const idf = idfCache.get(term) ?? 0.0;
      const infoQi = infoWeights.get(term) ?? 0.0;

      // Term-adaptive γt: rare terms get full entropy weight,
      // common terms get none — corpus-size independent
      const gammaT = idf * invIdfMax;
      const infoXQ = gammaT * infoQi * qTf;

      if (idf <= 0.0) {
        for (const cid of posting.keys()) {
          tanhCoverage.set(cid, (tanhCoverage.get(cid) ?? 0.0) + 1.0);
          infoAccum.set(cid, (infoAccum.get(cid) ?? 0.0) + infoXQ);
        }
        continue;
      }

      // Entropy-aware IDF with term-adaptive scaling
      const eidf = idf * (1.0 + gammaT * infoQi);

      for (const [cid, tf] of posting.entries()) {
        // BM25 TF saturation — proven, well-calibrated
        const dl = docLengths.get(cid)!;
        const tfSat = (tf * (k1 + 1.0)) / (tf + k1 * (1.0 - b + b * dl / avgdl));

        const termScore = eidf * tfSat * qTf;
        scores.set(cid, (scores.get(cid) ?? 0.0) + termScore);
        infoAccum.set(cid, (infoAccum.get(cid) ?? 0.0) + infoXQ);
        tanhCoverage.set(cid, (tanhCoverage.get(cid) ?? 0.0) + Math.tanh(termScore));
      }
    }

    if (scores.size === 0) return [];

    // Final: Soft-AND coverage bonus
    const finalScores: Array<[string, number]> = [];
    for (const [cid, base] of scores.entries()) {
      const softAnd = (tanhCoverage.get(cid) ?? 0.0) * invM;
      finalScores.push([cid, base + softAnd * (infoAccum.get(cid) ?? 0.0)]);
    }

    // Sort descending by score
    finalScores.sort((a, b_) => b_[1] - a[1]);

    if (this.normalizeScores && finalScores.length > 0) {
      const maxScore = finalScores[0]![1];
      if (maxScore > 0.0) {
        for (let i = 0; i < finalScores.length; i++) {
          finalScores[i] = [finalScores[i]![0], finalScores[i]![1] / maxScore];
        }
      }
    }

    return finalScores.slice(0, topK);
  }
}
